jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -77,68 +92,103 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3LazyLoa
-                    bucketName,
-                    this,
-                    "${this.key}/${message("s3.load_more")}",
-                    (nextPage.last() as S3TreeObjectVersionNode).versionId
-                )
+            val response = runBlocking {
+                bucket.listObjectVersions(key, continuationMarker)
+            }
+
+            return mutableListOf<S3TreeNode>().apply {
+                response?.versions()
+                    ?.filter { it.key() == key && it.versionId() != NOT_VERSIONED_VERSION_ID }
+                    ?.map { S3TreeObjectVersionNode(this@S3TreeObjectNode, it.versionId(), it.size(), it.lastModified()) }
+                    ?.onEach { add(it) }
+
+                if (response?.isTruncated == true) {
+                    response.versionIdMarker()?.let {
+                        S3TreeContinuationNode(bucket, this@S3TreeObjectNode, this@S3TreeObjectNode.key, it)
+                    }?.let {
+                        add(it)
+                    }
+                }
             }
-            return nextPage
         }
+
         return emptyList()
     }
 }
 
-class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3TreeObjectNode, key: String, size: Long, lastModified: Instant, val versionId: String) :
-    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
+class S3TreeObjectVersionNode(parent: S3TreeObjectNode, override val versionId: String, override val size: Long, override val lastModified: Instant) :
+    S3TreeNode(parent.bucket, parent, parent.key), S3Object {
+
+    init {
+        icon = parent.icon
+    }
+
+    override fun directoryPath(): String = (parent as S3TreeObjectNode).directoryPath()
 
-    override fun getName(): String {
-        // For not versioned buckets api can return versionId as literal 'null' so we avoid propagating null string to UI.
-        val versionId = if (versionId != NOT_VERSIONED_VERSION_ID) versionId else (parent as S3TreeObjectNode).name
-        val originalExtension = FileUtilRt.getExtension((parent as S3TreeObjectNode).name)
-        val extension = if (originalExtension.isNotBlank()) ".$originalExtension" else ""
+    override fun fileName(): String {
+        val parentObjectName = (parent as S3TreeObjectNode).fileName()
 
-        return "$versionId$extension"
+        val filenamePrefix = FileUtilRt.getNameWithoutExtension(parentObjectName) + "@" + versionId
+        val extension = FileUtilRt.getExtension(parentObjectName)
+        return if (extension.isNotEmpty()) {
+            "$filenamePrefix.$extension"
+        } else {
+            filenamePrefix
+        }
     }
 
+    override fun displayName(): String = versionId
+
     override fun getChildren(): Array<S3TreeNode> = emptyArray()
+
+    override fun getEqualityObjects(): Array<Any> = arrayOf(bucket, key, versionId)
+
+    override fun toString(): String = "S3TreeObjectVersionNode(key='$key', versionId='$versionId')"
```

Comment:
isnt the super impl do the same thing here?

Reply:
Super lacks versionId